### PR TITLE
Subscription install improvements

### PR DESCRIPTION
**Pull request by bgrgicak**
### Changes proposed in this Pull Request:

This PR fixes a few edge case issues with subscription installations by catching errors and not interrupting ongoing installations. 

The install process on the frontend will now run until the subscription list is updated.

### How to test the changes in this Pull Request:

Using the [WooCommerce Testing Instructions Guide](https://github.com/woocommerce/woocommerce/wiki/Writing-high-quality-testing-instructions), include your detailed testing instructions:

1. Open the [new subscriptions page](http://localhost:8888/wp-admin/admin.php?page=wc-admin&tab=my-subscriptions&path=%2Fextensions)
2. Install a product
3. [Remove the product](http://localhost:8888/wp-admin/plugins.php)
4. Install it again 
5. All steps should work as expected

### Changelog entry

-   [x] Automatically create a changelog entry from the details below.

<details>

#### Significance

-   [x] Patch
-   [ ] Minor
-   [ ] Major

#### Type

-   [x] Fix - Fixes an existing bug
-   [ ] Add - Adds functionality
-   [ ] Update - Update existing functionality
-   [ ] Dev - Development related task
-   [ ] Tweak - A minor adjustment to the codebase
-   [ ] Performance - Address performance issues
-   [ ] Enhancement - Improvement to existing functionality

#### Comment 

Catch subscription installation errors and wait for the updated subscription list to load.

</details>

### CODE REVIEW

**Review by andfinally — COMMENTED** [review]
When I try to install DePay, the request to `http://localhost:8888/wp-json/wc/v3/marketplace/subscriptions/install?_locale=user` gets 500. The error is

```[19-Oct-2023 10:04:39 UTC] PHP Fatal error:  Uncaught Error: Call to a member function get_last_step_status() on null in /var/www/html/wp-content/plugins/woocommerce/includes/admin/helper/class-wc-helper-subscriptions-api.php:238
Stack trace:
#0 /var/www/html/wp-includes/rest-api/class-wp-rest-server.php(1194): WC_Helper_Subscriptions_API::install(Object(WP_REST_Request))
#1 /var/www/html/wp-includes/rest-api/class-wp-rest-server.php(1041): WP_REST_Server->respond_to_request(Object(WP_REST_Request), '/wc/v3/marketpl...', Array, NULL)
#2 /var/www/html/wp-includes/rest-api/class-wp-rest-server.php(431): WP_REST_Server->dispatch(Object(WP_REST_Request))
#3 /var/www/html/wp-includes/rest-api.php(418): WP_REST_Server->serve_request('/wc/v3/marketpl...')
#4 /var/www/html/wp-includes/class-wp-hook.php(310): rest_api_loaded(Object(WP))
#5 /var/www/html/wp-includes/class-wp-hook.php(334): WP_Hook->apply_filters(NULL, Array)
#6 /var/www/html/wp-includes/plugin.php(565): WP_Hook->do_action(Array)
#7 /var/www/html/wp-includes/class-wp.php(398): do_action_ref_array( in /var/www/html/wp-content/plugins/woocommerce/includes/admin/helper/class-wc-helper-subscriptions-api.php on line 238
```

So it looks like the product state we get in `WC_WCCOM_Site_Installation_State_Storage` is null. I get the same thing when I try to install Product Recommendations.

I took the liberty of pushing a small CSS change in [`a6104f1` (#40854)](https://github.com/woocommerce/woocommerce/pull/40854/commits/a6104f1b2f1866a30d9c9abce9b75e45dee64f59) to fix the WordPress notifications issue 18448-gh-Automattic/woocommerce.com.

**Comment by bgrgicak** [comment]
@andfinally can you try again? It should be fixed now.

**Comment by andfinally** [comment]
Thanks @bgrgicak, I'm getting a 200 response with `The extension has been installed.` from the POST request to `/wp-json/wc/v3/marketplace/subscriptions/install`. But the subsequent request to `/wp-json/wc/v3/marketplace/subscriptions` gets DePay with `local.installed = false` and `local.active = false`, and on refreshing the page DePay is still not installed.

**Comment by bgrgicak** [comment]
@andfinally there is another PR that will address this https://github.com/woocommerce/woocommerce/pull/40820

File: plugins/woocommerce-admin/client/marketplace/components/my-subscriptions/table/rows/install.tsx [hunk 1]
```diff
@@ -69,6 +69,10 @@ export default function Install( props: ActivationToggleProps ) {
 		return __( 'Install', 'woocommerce' );
 	}
 
+	if ( props.subscription.local.installed === true ) {
```
Comment (andfinally) [inline plugins/woocommerce-admin/client/marketplace/components/my-subscriptions/table/rows/install.tsx:72]: I assume from this there will always be a `local` property.

**Review by andfinally — APPROVED** [review]
Thanks, works nicely! ❤️ 

I can't remember from our discussions with Poli whether we're still going to offer some way to toggle activated/connected. Presumably we're addressing that separately.

File: plugins/woocommerce-admin/client/marketplace/components/my-subscriptions/table/rows/install.tsx (hunk 1 above)
Comment (bgrgicak) [reply plugins/woocommerce-admin/client/marketplace/components/my-subscriptions/table/rows/install.tsx:72]: Yes, it's always added by the API. https://github.com/woocommerce/woocommerce/blob/345234dcabf8429c8e276340e347d17ae95b64e0/plugins/woocommerce/includes/admin/helper/class-wc-helper.php#L1463

**Comment by andfinally** [comment]
@bgrgicak Just been looking at the [thread](https://github.com/woocommerce/woocommerce/pull/40626#discussion_r1364259357) in this other PR, and it seems this PR doesn't change the idempotency key business, is that correct?

**Comment by bgrgicak** [comment]
>  it seems this PR doesn't change the idempotency key 

It doesn't change it. Looking at Denis' comment we should make it unique.
I don't think that we need this endpoint anymore. If we decide to move to the WP install as we discussed, we can delete it and avoid issues.

**Pull request closed without merging**